service_configuration_lib/spark_config.py

```diff
@@ -470,30 +486,147 @@ def compute_executor_instances_k8s(user_spark_opts: Dict[str, str]) -> int:
+                f'  - spark.executor.cores:     {cpu:3}c  → {new_cpu}c\n'
+                f'  - spark.executor.memory:    {memory:3}g  → {new_memory}g\n'
+                f'  - spark.executor.instances: {instances:3}x  → {new_instances}x\n'
+                'Check y/spark-metrics to compare how your job performance compared to previous runs.\n'
+                'Feel free to adjust to spark resource configs in yelpsoa-configs with above newly adjusted values.\n',
+            )
+
+        if new_cpu < task_cpus:
+            log.warning(
+                f'Given spark.task.cpus is {task_cpus}, '
+                f'=> adjusted to {new_cpu} to keep it within the limits of adjust spark.executor.cores.\n',
+            )
+            task_cpus = new_cpu
+        return new_cpu, f'{new_memory}g', new_instances, task_cpus
+
+    # Constants
+    recommended_memory_gb = RECOMMENDED_RESOURCE_CONFIGS['recommended'].executor_memory
+    medium_cores = RECOMMENDED_RESOURCE_CONFIGS['medium'].executor_cores
+    medium_memory_mb = RECOMMENDED_RESOURCE_CONFIGS['medium'].executor_memory
+    max_cores = RECOMMENDED_RESOURCE_CONFIGS['max'].executor_cores
+    max_memory_gb = RECOMMENDED_RESOURCE_CONFIGS['max'].executor_memory
+
+    if memory_gb > max_memory_gb or executor_cores > max_cores:
+        executor_cores, executor_memory = _cap_executor_resources(executor_cores, executor_memory, memory_mb)
+    elif force_spark_resource_configs:
         log.warning(
-            f'Given executor cores is {executor_cores}, '
-            f'capped to {EXECUTOR_CORES_CAP} to better fit on available aws nodes.',
+            '--force-spark-resource-configs is set to true: '
+            'this can result in non-optimal bin-packing of executors on aws nodes or '
+            'can lead to wastage the resources. '
+            "Please use this flag only if you have tested that standard memory/cpu configs won't work for your job.\n"
+            'Let us know at #spark if you think, your use-case needs to be standardized.\n',
+        )
+    elif memory_gb > medium_memory_mb or executor_cores > medium_cores:
+        (executor_cores, executor_memory, executor_instances, task_cpus) = _calculate_resources(
+            executor_cores,
+            memory_gb,
+            executor_instances,
+            task_cpus,
+            medium_memory_mb,
+            ratio_adj_thresh,
+        )
+    else:
+        (executor_cores, executor_memory, executor_instances, task_cpus) = _calculate_resources(
+            executor_cores,
+            memory_gb,
+            executor_instances,
+            task_cpus,
+            recommended_memory_gb,
+            ratio_adj_thresh,
         )
-        executor_cores = EXECUTOR_CORES_CAP
 
     user_spark_opts.update({
-        'spark.executor.memory': str(executor_memory),
         'spark.executor.cores': str(executor_cores),
+        'spark.executor.memory': str(executor_memory),
+        'spark.executor.instances': str(executor_instances),
```

Comment:
How would this work with dynamicAllocation? You have changed number of instances but there is no change in the maxExecutors for DRA

Reply:
I think DRA configs will be set later in `get_dra_configs()` according to instance count after this function?

Reply:
Would it makes sense to test or add 1 test case to observe how the configs adapt when executor instances are specified as part of DRA configs but not the standard `spark.executor.instances`?